archive_viewer/table_models/curve_model.py

```diff
@@ -96,16 +121,107 @@ def append(self, address: Optional[str] = None, name: Optional[str] = None, colo
+        self._plot._curves[index.row()] = self._plot.replaceToArchivePlot(address=address, name=address, color=color, yAxisName=y_axis.name)
+        self.plot.linkDataToAxis(self._plot._curves[index.row()], y_axis.name)
+        del curve
+    def replaceToFormula(self, index: QModelIndex, formula:str, color: Optional[QColor] = None) -> bool:
+        """Add a new curve item to plot and the data model.
+
+        Parameters
+        ----------
+        address : str, optional
+            The PV address that the curve should gather data from.
+        name : str, optional
+            The display name for the curve.
+        color : Optional[QColor], optional
+            The curve's color on the plot.
+        """
+        #First, find all the rows by regexing for the {} format
+        pvs = re.findall("{(.+?)}", formula)
+        pvdict = dict()
+        rowName = self._row_names[index.row()]
+        for pv in pvs:
+            #Check if all of the requested rows actually exist
+            if pv not in self._row_names:
+                print("Error, " + pv + " is an invalid variable name.")
+                return False
+            elif pv == rowName:
+                print("Error, formula is recursive")
+                return False
+            else:
+                #if it's good, add it to the dictionary of curves. rindex = row index (int) as opposed to index, which is a QModelIndex
+                rindex = self._row_names.index(pv)
+                pvdict[pv] = self._plot._curves[rindex]
+        print("accepted formula")
+        curve = self._plot._curves[index.row()] 
+        if not color:
+            color = ColorButton.index_color(index.row())
+        #          KLYS:LI22:31:KVAC
+        #Handle Archives and formulas differently
+        if isinstance(curve, ArchivePlotCurveItem):
+            if index.row() == self.rowCount() - 1:
+                self._axis_model.append()
+                self.append()
+            y_axis = self._axis_model.get_axis(index.row())
+            self._plot._curves[index.row()] = self._plot.addFormulaChannel(formula=formula, name=formula, pvs=pvdict,color=color, useArchiveData=True, yAxisName=y_axis.name)
+            self._plot._curves[index.row()].formula_invalid_signal.connect(partial(self.invalidFormula, header = rowName))
+            #Need to check if Formula is referencing a dead row
+            self.plot.plotItem.unlinkDataFromAxis(curve.y_axis_name)
+            self.plot.removeItem(curve)
+            #Disconnect everything and delete it, create a new Formula with the dictionary of curve
+            [ch.disconnect() for ch in curve.channels() if ch]
+            del curve
+        else:
+            #if we already were a formula, just use the new formula data
+            curve.formula = formula
+            curve.pvs = pvdict
+        #redraw
+        self._plot.set_needs_redraw()
+        self._plot.redrawPlot()
+        return True
+    
+    def invalidFormula(self, header):
```

Comment:
No because we don't want to change the list of items. We want to edit the curve in place - deleting the curve with removeAtIndex also destroys its position in the list, which I guess we could fix by inserting but I'm not sure if that's better

Reply:
Sounds good. I'll check this in PR #25